src/lib/common/include/sol-mainloop.h

```diff
@@ -64,11 +64,269 @@ extern "C" {
  * @li @ref Update
  * @li @ref Utils
  * @li @ref WorkerThread
+ * @li @ref patterns
  *
  * Please see the @ref authors page for contact details.
  */
 
 /**
+ * @page patterns Design Patterns
+ * @tableofcontents
+ *
+ * This page will explain the common design patterns employed by Soletta.
```

Comment:
the documentation is good but this is the "implementation side", good for people that develop soletta itself or those that want to make compatible API.

There is a more important case that is missing: our users :-) That one can be in a different section (or move the current text to https://github.com/solettaproject/soletta/wiki/API-design-conventions and only document the user usage here. Cases to explicitly address:
 - simple `rx_cb` and `tx_cb` usages
 - delete from callbacks (say it's allowed)
 - reading partial contents (ie: return `< buf->used` from `rx_cb` is allowed), say what can happen if reaches `rx_size` (you can relate that to `buf->capacity`, say that in that case user must make room for more data, either by keeping a local copy or discard, in both cases return `> 0` to allow more data to be read.
 - saturating the write channel (ie: `returns -ENOSPC` with a `tx_size`), you should control your pipeline to stop producing data.